### PR TITLE
Move from `max( id )` to `max( index )` for latest commit statuses

### LINKED ISSUE #30074: Gitea shows outdated commit status with distributed DB backend (eg. TiDB)
### Description

Gitea doesn’t always show the latest commit status (i.e. being outdated) if you use some distributed database as backend. An `auto_increment` primary key, though being expected to bump at a step of 1 at every `INSERT`, may be designed to bump independently on each nodes.

We should update `commit_status.go` to use the ``(max(`index`), repo_id, sha)`` triple to determine the latest commit status, instead of relying on `max(id)`.

### Gitea Version

1.21.9

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

I’m running Gitea on a Kubernetes cluster, which also hosts Woodpecker CI v2.3.0 and TiDB v7.5.1. Gitea uses TiDB (through `mysql`) as backend and Woodpecker uses Gitea as forge.

The problem was originally identified as “randomly dropping commit status on either Woodpecker, Gitea and TiDB side”, but after digging into the database it turns out that all commit statuses were there.

### Database

MySQL/MariaDB

### PR DESCRIPTION
**Pull request by stevapple**
This PR replaces the use of `max( id )`, and instead using ``max( `index` )`` for determining the latest commit status. Building business logic over an `auto_increment` primary key like `id` is risky and there’re already plenty of discussions on the Internet.

There‘s no guarantee for `auto_increment` values to be monotonic, especially upon failures or with a cluster. In the specific case, we met the problem of commit statuses being outdated when using TiDB as the database. As [being documented](https://docs.pingcap.com/tidb/stable/auto-increment), `auto_increment` values assigned to an `insert` statement will only be monotonic on a per server (node) basis.

Closes #30074.

### CODE REVIEW

**Comment by stevapple** [comment]
Ping for a review

**Review by wxiaoguang — APPROVED** [review]

File: models/git/commit_status.go [hunk 1]
```diff
@@ -269,44 +270,48 @@ type CommitStatusIndex struct {
...
 
 	// Create a disjunction of conditions for each repoID and SHA pair
 	conds := make([]builder.Cond, 0, len(repoIDsToLatestCommitSHAs))
 	for repoID, sha := range repoIDsToLatestCommitSHAs {
 		conds = append(conds, builder.Eq{"repo_id": repoID, "sha": sha})
 	}
-	sess = sess.Where(builder.Or(conds...)).
-		Select("max( id ) as id, repo_id").
-		GroupBy("context_hash, repo_id").OrderBy("max( id ) desc")
+	sess := getBase().Where(builder.Or(conds...)).
```
Comment (lunny) [inline models/git/commit_status.go:312]: Maybe we can get these indexes from table `commit_status_index` directly, then we don't need to use `max`.

Comment (stevapple) [reply models/git/commit_status.go:312]: Records in `commit_status_index` are grouped by `repo_id` and `sha`, but not `context_hash`. It doesn’t seem helpful for the cases.

**Review by lunny — APPROVED** [review]

**Comment by stevapple** [comment]
@lunny May it be back-ported to v1.21, or is v1.22 around the corner?

**Pull request merged**